### PR TITLE
Feature/refactoring step 2

### PR DESCRIPTION
**Pull request by anthonyGuiguen**

### CODE REVIEW

File: ZetaPushSwift/ZetaPushMacroListener.swift [hunk 1]
```diff
@@ -1,205 +1,202 @@
...
+    self.init(clientHelper, deploymentId: zetaPushDefaultConfig.macroDeployementId)
+  }
+  /**
+   
+   */
+  public func getModelBlock<T: Glossy>(verb: String, callback: @escaping (T) -> Void) -> ModelBlockTuple {
+    let channel = self.clientHelper.composeServiceChannel(verb, deploymentId: zetaPushMacroService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { [weak self] (messageDict: NSDictionary) -> Void in
+      guard let self = self else { return }
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushMacroListener.swift:39]: add log here

File: ZetaPushSwift/ZetaPushMacroListener.swift [hunk 2]
```diff
@@ -1,205 +1,202 @@
...
+      }
+      callback(zpMessage)
+    })
+  }
+  
+  public func getModelBlock<T: Glossy>(verb: String, callback: @escaping ([T]) -> Void) -> ModelBlockTuple {
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushMacroService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { [weak self] (messageDict: NSDictionary) -> Void in
+      guard let self = self else { return }
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushMacroListener.swift:53]: add log here

File: ZetaPushSwift/ZetaPushMacroListener.swift [hunk 3]
```diff
@@ -1,205 +1,202 @@
...
+      }
+      callback(zpMessage)
+    })
+  }
+  
+  public func getModelBlock<T: AbstractMacroCompletion>(verb: String, callback: @escaping (T) -> Void) -> ModelBlockTuple {
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushMacroService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { [weak self] (messageDict: NSDictionary) -> Void in
+      guard let self = self else { return }
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushMacroListener.swift:67]: add log here

File: ZetaPushSwift/ZetaPushMacroListener.swift [hunk 4]
```diff
@@ -1,205 +1,202 @@
...
+      }
+      callback(zpMessage)
+    })
+  }
+  
+  public func getModelBlock<T: NSDictionary>(verb: String, callback: @escaping (T) -> Void) -> ModelBlockTuple {
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushMacroService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { [weak self] (messageDict: NSDictionary) -> Void in
+      guard let self = self else { return }
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushMacroListener.swift:81]: add log here

File: ZetaPushSwift/ZetaPushMacroListener.swift [hunk 5]
```diff
@@ -1,205 +1,202 @@
...
-        }
-        let channel: String = self.clientHelper.composeServiceChannel(verb, deploymentId: self.zetaPushMacroService.deploymentId)
-        self.clientHelper.subscribe(channel, block: channelBlockServiceCall)
-        
+  }
+  
+  /// Generic Subscribe with a NSDictionary parameter
+  public func subscribe<T: NSDictionary>(verb: String, callback: @escaping (T) -> Void) {
+    let channelBlockServiceCall:ChannelSubscriptionBlock = { [weak self] (messageDict) -> Void in
+      guard let self = self else { return }
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushMacroListener.swift:145]: add log here

File: ZetaPushSwift/ZetaPushMacroListener.swift [hunk 6]
```diff
@@ -1,205 +1,202 @@
...
-        return zpMessage
-    }
+    let channel = self.clientHelper.composeServiceChannel(verb, deploymentId: zetaPushMacroService.deploymentId)
+    self.clientHelper.subscribe(channel, block: channelBlockServiceCall)
+  }
+  
+  // MARK: - private funcs
+  private func handleMacroErrors(from messageDict: NSDictionary) {
+    guard let errors = messageDict["errors"] as? [Any], !errors.isEmpty else {
+      return
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushMacroListener.swift:161]: add log here

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 7]
```diff
@@ -1,140 +1,141 @@
...
+open class ZetaPushServiceListener {
+  // MARK: Properties
+  public let clientHelper: ClientHelper
+  var macroChannelError: String
+  public var zetaPushService: ZetaPushService
+  open var onServiceError: ZPServiceErrorBlock?
+  
+  // Callback for /error macro channel
+  lazy var channelBlockMacroError:ChannelSubscriptionBlock = { [weak self] (messageDict) -> Void in
+    guard let self = self else { return }
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift:28]: add log here

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 8]
```diff
@@ -1,140 +1,141 @@
...
-    public func subscribe(_ tuples: [ModelBlockTuple]) {
-        self.clientHelper.subscribe(tuples)
+    self.register()
+  }
+  
+  ///
+  public func getModelBlock<T: Glossy>(verb: String, callback: @escaping (T) -> Void) -> ModelBlockTuple {
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { (messageDict: NSDictionary) -> Void in
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift]: add weak self here and don't forget to log ;)

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 9]
```diff
@@ -1,140 +1,141 @@
...
+      }
+      callback(zpMessage)
+    })
+  }
+  
+  ///
+  public func getModelBlock<T: Glossy>(verb: String, callback: @escaping ([T]) -> Void) -> ModelBlockTuple {
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { (messageDict: NSDictionary) -> Void in
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift]: add weak self here and don't forget to log ;)

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 10]
```diff
@@ -1,140 +1,141 @@
...
+      }
+      callback(zpMessage)
+    })
+  }
+  
+  ///
+  public func getModelBlock<T: NSDictionary>(verb: String, callback: @escaping (T) -> Void) -> ModelBlockTuple {
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushService.deploymentId)
+    let model = CometdSubscriptionModel(subscriptionUrl: channel, clientId: self.clientHelper.cometdClient.cometdClientId)
+    return ModelBlockTuple(model: model, block: { (messageDict: NSDictionary) -> Void in
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift]: add weak self here and don't forget to log ;)

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 11]
```diff
@@ -1,140 +1,141 @@
...
+  }
+  
+  ///
+  public func subscribe(_ tuples: [ModelBlockTuple]) {
+    clientHelper.subscribe(tuples)
+  }
+  
+  /// Generic Subscribe with a Generic parameter
+  public func subscribe<T: Glossy>(verb: String, callback: @escaping (T) -> Void) {
+    let channelBlockServiceCall: ChannelSubscriptionBlock =  { (messageDict) -> Void in
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift]: add weak self here and don't forget to log ;)

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 12]
```diff
@@ -1,140 +1,141 @@
...
-        }
-        let channel: String = self.clientHelper.composeServiceChannel(verb, deploymentId: self.zetaPushService.deploymentId)
-        self.clientHelper.subscribe(channel, block: channelBlockServiceCall)
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushService.deploymentId)
+    clientHelper.subscribe(channel, block: channelBlockServiceCall)
+  }
+  
+  /// Generic Subscribe with a Generic Array parameter
+  public func subscribe<T: Glossy>(verb: String, callback: @escaping ([T]) -> Void) {
+    let channelBlockServiceCall: ChannelSubscriptionBlock = { (messageDict) -> Void in
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift]: add weak self here and don't forget to log ;)

File: ZetaPushSwift/ZetaPushServiceListener.swift [hunk 13]
```diff
@@ -1,140 +1,141 @@
...
-    
-    private func parse<T: NSDictionary>(messageDict: NSDictionary) -> T? {
-        return messageDict as? T
+    let channel = clientHelper.composeServiceChannel(verb, deploymentId: zetaPushService.deploymentId)
+    clientHelper.subscribe(channel, block: channelBlockServiceCall)
+  }
+  
+  /// Generic Subscribe with a NSDictionary parameter
+  public func subscribe<T: NSDictionary>(verb: String, callback: @escaping (T) -> Void) {
+    let channelBlockServiceCall: ChannelSubscriptionBlock = { (messageDict) -> Void in
```
Comment (StevenWatremez) [inline ZetaPushSwift/ZetaPushServiceListener.swift]: add weak self here and don't forget to log ;)

**Review by StevenWatremez — CHANGES_REQUESTED** [review]

File: ZetaPushSwift/ZetaPushMacroListener.swift (hunk 1 above)
Comment (anthonyGuiguen) [reply ZetaPushSwift/ZetaPushMacroListener.swift:39]: We'll do that in a special Log issue

File: ZetaPushSwift/ZetaPushMacroListener.swift (hunk 2 above)
Comment (anthonyGuiguen) [reply ZetaPushSwift/ZetaPushMacroListener.swift:53]: We'll do that in a special Log issue

File: ZetaPushSwift/ZetaPushMacroListener.swift (hunk 3 above)
Comment (anthonyGuiguen) [reply ZetaPushSwift/ZetaPushMacroListener.swift:67]: We'll do that in a special Log issue

File: ZetaPushSwift/ZetaPushMacroListener.swift (hunk 4 above)
Comment (anthonyGuiguen) [reply ZetaPushSwift/ZetaPushMacroListener.swift:81]: We'll do that in a special Log issue

File: ZetaPushSwift/ZetaPushMacroListener.swift (hunk 5 above)
Comment (anthonyGuiguen) [reply ZetaPushSwift/ZetaPushMacroListener.swift:145]: We'll do that in a special Log issue

File: ZetaPushSwift/ZetaPushMacroListener.swift (hunk 6 above)
Comment (anthonyGuiguen) [reply ZetaPushSwift/ZetaPushMacroListener.swift:161]: We'll do that in a special Log issue

**Review by StevenWatremez — APPROVED** [review]

**Pull request merged**